v2/devdoc/logging.md

```diff
@@ -0,0 +1,339 @@
+# `c-logging` requirements
+
+## Basic functionality
+
+`c-logging` shall expose APIs allowing the user to log events that can be used for observing the software activity, debugging, etc.
+
+The exposed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    logger_log(log_level, log_context, __FILE__, __FUNCTION__, __LINE__, format MU_IFCOMMALOGIC(MU_COUNT_ARG(__VA_ARGS__)) __VA_ARGS__)
+
+#define LOGGER_LOG_EX(log_level, ...) \
+    ...
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack.
+// stack allocated contexts do not require any destructor
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
```

Comment:
>log_context [](http://example.com/codeflow?start=33&length=11)

I remember seeing these needing "&" operator. I might be wrong though.\


Hmmm... line 82 says i don't need it. IIIINTERESTING.


Oh wait. definigin it doesnt need it, but using it does it? I am ... suprised

Reply:
Fixed the usage in LOG_CONTEXT_LOCAL_DEFINE and clarified that when the stack context is used it has to be used with &